engine/src/main/java/io/zeebe/engine/state/EventApplier.java

```diff
@@ -0,0 +1,22 @@
+/*
+ * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH under
+ * one or more contributor license agreements. See the NOTICE file distributed
+ * with this work for additional information regarding copyright ownership.
+ * Licensed under the Zeebe Community License 1.0. You may not use this file
+ * except in compliance with the Zeebe Community License 1.0.
+ */
+package io.zeebe.engine.state;
+
+import io.zeebe.msgpack.UnpackedObject;
+import io.zeebe.protocol.impl.record.RecordMetadata;
+import io.zeebe.protocol.record.intent.Intent;
+
+/** Applies state changes for a specific event to the {@link io.zeebe.engine.state.ZeebeState}. */
+public interface EventApplier {
```

Comment:
One thing that makes me feel a little uneasy is that this interface fits both to an individual event applier as well as to the `EventAppliers` class. This often leads to confusion when the relay class has a similar interface to the delegate classes.

It also means that I can call the event applier for a particular event with any arbitrary intent + value type, and there are no safeguards against this.

One option to remedy this would be to make this interface more specific (e.g. remove the intent, because it is already predetermined)

If the interface should remain versatile as it is now. Another option is to turn this interface into an abstract class and implement safeguards into the abstract class.

```
abstract List<Intent> supportedIntents();
abstract List<ValueType> supportedValueTypes();

public void applyState(...) {
   if (!supportedIntents.contains(intent)) {
    throw new IllegalArgumentException("not responsilble for this");
   }

  applyStateInternal(...)
```

Reply:
We've found some solutions to improve this